akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -79,24 +79,38 @@ import com.github.ghik.silencer.silent
     new GraphStageLogic(shape) with OutHandler with InHandler {
       def decider = inheritedAttributes.mandatoryAttribute[SupervisionStrategy].decider
 
-      override def onPush(): Unit = {
+      private var buffer: OptionVal[T] = OptionVal.none
+
+      override def preStart(): Unit = pull(in)
+      override def onPush(): Unit =
         try {
           val elem = grab(in)
-          if (p(elem)) {
-            push(out, elem)
-          } else {
-            pull(in)
-          }
+          if (p(elem))
+            if (isAvailable(out)) {
+              push(out, elem)
+              tryPull(in)
+            } else
+              buffer = OptionVal.Some(elem)
+          else tryPull(in)
```

Comment:
can these `tryPull` be `pull` or is there some case where it's not possible to pull here?

Reply:
Yep, shouldn't be possible in onPush that the inlet is already completed.